Ring PutOne panics with unaligned 64-bit on Raspberry PI 4 (arm64)

Spotted while running GitLab-CE (for which their KAS component is using this Rueidis client) and it keeps streaming this panic error in logs

```bash
2023-11-11_17:52:14.60237 goroutine 67 [running]:
2023-11-11_17:52:14.60243 runtime/internal/atomic.panicUnaligned()
2023-11-11_17:52:14.60247       /usr/local/go/src/runtime/internal/atomic/unaligned.go:8 +0x24
2023-11-11_17:52:14.60250 runtime/internal/atomic.Xadd64(0x4149354, 0x1)
2023-11-11_17:52:14.60361       /usr/local/go/src/runtime/internal/atomic/atomic_arm.s:258 +0x14
2023-11-11_17:52:14.60365 github.com/redis/rueidis.(*ring).PutOne(0x4149320, {0x4012360, 0x3000, 0x8000})
2023-11-11_17:52:14.60372       /root/go/pkg/mod/github.com/redis/rueidis@v1.0.14/ring.go:55 +0x34
2023-11-11_17:52:14.60375 github.com/redis/rueidis.(*pipe).Do(0x3d76820, {0x131c490, 0x3deb560}, {0x4012360, 0x3000, 0x8000})
2023-11-11_17:52:14.60378       /root/go/pkg/mod/github.com/redis/rueidis@v1.0.14/pipe.go:814 +0x2ec
2023-11-11_17:52:14.60387 github.com/redis/rueidis.(*pipe).Receive(0x3d76820, {0x131c490, 0x3deb560}, {0x4012360, 0x3000, 0x8000}, 0x40f0030)
2023-11-11_17:52:14.60390       /root/go/pkg/mod/github.com/redis/rueidis@v1.0.14/pipe.go:687 +0x24c
2023-11-11_17:52:14.60393 github.com/redis/rueidis.(*mux).Receive(0x3d36000, {0x131c490, 0x3deb560}, {0x4012360, 0x3000, 0x8000}, 0x40f0030)
2023-11-11_17:52:14.60396       /root/go/pkg/mod/github.com/redis/rueidis@v1.0.14/mux.go:322 +0xc0
2023-11-11_17:52:14.60399 github.com/redis/rueidis.(*singleClient).Receive(0x3c0f890, {0x131c490, 0x3deb560}, {0x4012360, 0x3000, 0x8000}, 0x40f0030)
2023-11-11_17:52:14.60402       /root/go/pkg/mod/github.com/redis/rueidis@v1.0.14/client.go:107 +0x58
2023-11-11_17:52:14.60405 gitlab.com/gitlab-org/cluster-integration/gitlab-agent/v16/cmd/kas/kasapp.(*serverApi).subscribeToEvents.func1({0x131c490, 0x3deb560})
2023-11-11_17:52:14.60458       /var/cache/omnibus/src/gitlab-kas/cmd/kas/kasapp/server_api.go:103 +0x170
2023-11-11_17:52:14.60471 gitlab.com/gitlab-org/cluster-integration/gitlab-agent/v16/internal/tool/retry.PollWithBackoff({0x131c490, 0x3deb560}, {{0x1313ec8, 0x3ef08c0}, 0x2faf080, 0x1, 0x40ee0c0}, 0x3c2ef50)
2023-11-11_17:52:14.60475       /var/cache/omnibus/src/gitlab-kas/internal/tool/retry/retry.go:80 +0x238
2023-11-11_17:52:14.60478 gitlab.com/gitlab-org/cluster-integration/gitlab-agent/v16/cmd/kas/kasapp.(*serverApi).subscribeToEvents(0x4160b40, {0x131c490, 0x3deb560})
2023-11-11_17:52:14.60481       /var/cache/omnibus/src/gitlab-kas/cmd/kas/kasapp/server_api.go:101 +0x74
2023-11-11_17:52:14.60484 gitlab.com/gitlab-org/cluster-integration/gitlab-agent/v16/internal/module/notifications/server.(*module).Run(0x3fc1048, {0x131c490, 0x3deb560})
2023-11-11_17:52:14.60487       /var/cache/omnibus/src/gitlab-kas/internal/module/notifications/server/module.go:16 +0x30
2023-11-11_17:52:14.60490 gitlab.com/gitlab-org/cluster-integration/gitlab-agent/v16/cmd/kas/kasapp.startModules.func1({0x131c490, 0x3deb560})
2023-11-11_17:52:14.60493       /var/cache/omnibus/src/gitlab-kas/cmd/kas/kasapp/configured_app.go:663 +0x3c
2023-11-11_17:52:14.60497 github.com/ash2k/stager.(*stage).Go.func1()
2023-11-11_17:52:14.60500       /root/go/pkg/mod/github.com/ash2k/stager@v0.3.0/stage.go:25 +0x34
2023-11-11_17:52:14.60503 created by github.com/ash2k/stager.(*stage).Go
2023-11-11_17:52:14.60522       /root/go/pkg/mod/github.com/ash2k/stager@v0.3.0/stage.go:24 +0x7c
2023-11-11_17:52:15.74897 panic: unaligned 64-bit atomic operation
```

I reproduce the issue when cloning redis/rueidis and running Ring unit test:

```bash
go test -v -run TestRing

=== RUN   TestRing
=== RUN   TestRing/PutOne
panic: unaligned 64-bit atomic operation

goroutine 38 [running]:
runtime/internal/atomic.panicUnaligned()
        /usr/local/go/src/runtime/internal/atomic/unaligned.go:8 +0x24
runtime/internal/atomic.Xadd64(0x25244b4, 0x1)
        /usr/local/go/src/runtime/internal/atomic/atomic_arm.s:258 +0x14
github.com/redis/rueidis.(*ring).PutOne(0x2524480, {0x2564588, 0x0, 0x0})
        /home/xavier/go/src/rueidis/ring.go:55 +0x34
created by github.com/redis/rueidis.TestRing.func1 in goroutine 35
        /home/xavier/go/src/rueidis/ring_test.go:24 +0xf8
panic: unaligned 64-bit atomic operation
[...]
```

My hardware is a Raspberry PI 4, running [Raspberry Pi OS (64-bit)](https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-64-bit), I've installed Go using standard instructions and [go1.21.4.linux-arm64.tar.gz](https://go.dev/dl/go1.21.4.linux-arm64.tar.gz), however for some reason Go sets GOARCH to "arm" (ie, 32bit) instead of "arm64" - see `go env` output below:

Distro:
```
uname -a
Linux raspberrypi 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux
```
```
cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

Go env:
```
go version
go version go1.21.4 linux/arm64
```
```
go env
GO111MODULE=''
GOARCH='arm'
GOBIN=''
GOCACHE='/home/xavier/.cache/go-build'
GOENV='/home/xavier/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/xavier/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/xavier/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_arm64'
GOVCS=''
GOVERSION='go1.21.4'
GCCGO='gccgo'
GOARM='7'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='0'
GOMOD='/home/xavier/go/src/rueidis/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -marm -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build2790280466=/tmp/go-build -gno-record-gcc-switches'
```

Googling for it, I found this that may be a clue / starting point: https://github.com/golang/go/issues/23345

From what I was able to tell by running go test on the whole project, the atomic.AddUint64 from lru.Flight() are also panicing similarly, ex: 

```
go test -v .
=== RUN   TestCacheStore
=== RUN   TestCacheStore/LRUCacheStore
=== RUN   TestCacheStore/LRUCacheStore/Flight_and_Update
--- FAIL: TestCacheStore (0.00s)
    --- FAIL: TestCacheStore/LRUCacheStore (0.00s)
        --- FAIL: TestCacheStore/LRUCacheStore/Flight_and_Update (0.00s)
panic: unaligned 64-bit atomic operation [recovered]
        panic: unaligned 64-bit atomic operation

goroutine 25 [running]:
testing.tRunner.func1.2({0x422f00, 0x52eba0})
        /usr/local/go/src/testing/testing.go:1545 +0x27c
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1548 +0x42c
panic({0x422f00, 0x52eba0})
        /usr/local/go/src/runtime/panic.go:914 +0x244
runtime/internal/atomic.panicUnaligned()
        /usr/local/go/src/runtime/internal/atomic/unaligned.go:8 +0x24
runtime/internal/atomic.Xadd64(0x1c93234, 0x1)
        /usr/local/go/src/runtime/internal/atomic/atomic_arm.s:258 +0x14
github.com/redis/rueidis.(*lru).Flight(0x1ca6300, {0x46e6c6, 0x3}, {0x46e6c9, 0x3}, 0x5f5e100, {0xc14c5dcc1f58900a, 0xb4bd7b, 0x765ae0})
        /home/xavier/go/src/rueidis/lru.go:124 +0x76c
github.com/redis/rueidis.test.func1(0x1c9a870)
        /home/xavier/go/src/rueidis/cache_test.go:16 +0xa8
testing.tRunner(0x1c9a870, 0x1c8e4e8)
        /usr/local/go/src/testing/testing.go:1595 +0x124
created by testing.(*T).Run in goroutine 24
        /usr/local/go/src/testing/testing.go:1648 +0x3f0
FAIL    github.com/redis/rueidis        0.020s
FAIL
```

Also seen in these other tests (all with a smiliar call stack up to lru.Flight(..) )
- TestLRU/Cache_Hit_&_Expire
- TestClientSideCaching

At this stage I got lazy and stopped manually skipping KO testst to figure out the next one, instead I assumed it's unlikely some other path has issue and wasn't tested yet by all these tests. 

Also I tried running this ./docker-test.sh, but no luck, looks like some of the images needed by compose do not have arm64 versions:
```
./dockertest.sh

trap "docker compose down -v" EXIT
docker compose up -d
[+] Running 10/51
[...]
no matching manifest for linux/arm/v8 in the manifest list entries
docker compose down -v
```

My environment is still identical to the one I gave in my initial message (see https://github.com/redis/rueidis/issues/407#issue-1989078385)

Let me know if you'd like me to run some more tests or provide complementary info.

--- a/lru.go
+++ b/lru.go
@@ -20,7 +20,7 @@ const (
 	entryBaseSize = (keyCacheSize + entrySize + elementSize + stringSSize*2) * 3 / 2
 	entryMinSize  = entryBaseSize + messageStructSize
 
-	moveThreshold = uint64(1024 - 1)
+	moveThreshold = uint32(1024 - 1)
 )
 
 type cacheEntry struct {
@@ -48,8 +48,8 @@ func (e *cacheEntry) Wait(ctx context.Context) (RedisMessage, error) {
 type keyCache struct {
 	cache map[string]*list.Element
 	key   string
-	hits  uint64
-	miss  uint64
+	hits  uint32
+	miss  uint32
 }
 
 var _ CacheStore = (*lru)(nil)
@@ -87,7 +87,7 @@ func (c *lru) Flight(key, cmd string, ttl time.Duration, now time.Time) (v Redis
 	c.mu.RUnlock()
 
 	if e != nil && (v.typ == 0 || v.relativePTTL(now) > 0) {
-		hits := atomic.AddUint64(&kc.hits, 1)
+		hits := atomic.AddUint32(&kc.hits, 1)
 		if ele != back && hits&moveThreshold == 0 {
 			c.mu.Lock()
 			if c.list != nil {
@@ -111,7 +111,7 @@ func (c *lru) Flight(key, cmd string, ttl time.Duration, now time.Time) (v Redis
 	}
 	if ele, ok = kc.cache[cmd]; ok {
 		if e = ele.Value.(*cacheEntry); e.val.typ == 0 || e.val.relativePTTL(now) > 0 {
-			atomic.AddUint64(&kc.hits, 1)
+			atomic.AddUint32(&kc.hits, 1)
 			v = e.val
 			c.list.MoveToBack(ele)
 			ce = e
@@ -121,7 +121,7 @@ func (c *lru) Flight(key, cmd string, ttl time.Duration, now time.Time) (v Redis
 			c.size -= e.size
 		}
 	}
-	atomic.AddUint64(&kc.miss, 1)
+	atomic.AddUint32(&kc.miss, 1)
 	v.setExpireAt(now.Add(ttl).UnixMilli())
 	c.list.PushBack(&cacheEntry{cmd: cmd, kc: kc, val: v, ch: make(chan struct{})})
 	kc.cache[cmd] = c.list.Back()
@@ -147,7 +147,7 @@ func (c *lru) Flights(now time.Time, multi []CacheableTTL, results []RedisResult
 				} else {
 					goto miss1
 				}
-				if atomic.AddUint64(&kc.hits, 1)&moveThreshold == 0 {
+				if atomic.AddUint32(&kc.hits, 1)&moveThreshold == 0 {
 					if moves == nil {
 						moves = make([]*list.Element, 0, len(multi))
 					}
@@ -203,12 +203,12 @@ func (c *lru) Flights(now time.Time, multi []CacheableTTL, results []RedisResult
 				c.size -= e.size
 				goto miss2
 			}
-			atomic.AddUint64(&kc.hits, 1)
+			atomic.AddUint32(&kc.hits, 1)
 			c.list.MoveToBack(ele)
 			continue
 		}
 	miss2:
-		atomic.AddUint64(&kc.miss, 1)
+		atomic.AddUint32(&kc.miss, 1)
 		v := RedisMessage{}
 		v.setExpireAt(now.Add(multi[i].TTL).UnixMilli())
 		c.list.PushBack(&cacheEntry{cmd: cmd, kc: kc, val: v, ch: make(chan struct{})})

--- a/lru_test.go
+++ b/lru_test.go
@@ -78,16 +78,16 @@ func TestLRU(t *testing.T) {
 		store1 := lru.store["1"]
 		store2 := lru.store["2"]
 		lru.mu.RUnlock()
-		if miss := atomic.LoadUint64(&store1.miss); miss != 1 {
+		if miss := atomic.LoadUint32(&store1.miss); miss != 1 {
 			t.Fatalf("unexpected miss count %v", miss)
 		}
-		if hits := atomic.LoadUint64(&store1.hits); hits != uint64(count-1) {
+		if hits := atomic.LoadUint32(&store1.hits); hits != uint32(count-1) {
 			t.Fatalf("unexpected hits count %v", hits)
 		}
-		if miss := atomic.LoadUint64(&store2.miss); miss != 1 {
+		if miss := atomic.LoadUint32(&store2.miss); miss != 1 {
 			t.Fatalf("unexpected miss count %v", miss)
 		}
-		if hits := atomic.LoadUint64(&store2.hits); hits != uint64(count-1) {
+		if hits := atomic.LoadUint32(&store2.hits); hits != uint32(count-1) {
 			t.Fatalf("unexpected hits count %v", hits)
 		}
 	})
@@ -289,16 +289,16 @@ func TestLRU(t *testing.T) {
 		store1 := lru.store["1"]
 		store2 := lru.store["2"]
 		lru.mu.RUnlock()
-		if miss := atomic.LoadUint64(&store1.miss); miss != 1 {
+		if miss := atomic.LoadUint32(&store1.miss); miss != 1 {
 			t.Fatalf("unexpected miss count %v", miss)
 		}
-		if hits := atomic.LoadUint64(&store1.hits); hits != uint64(count-1) {
+		if hits := atomic.LoadUint32(&store1.hits); hits != uint32(count-1) {
 			t.Fatalf("unexpected hits count %v", hits)
 		}
-		if miss := atomic.LoadUint64(&store2.miss); miss != 1 {
+		if miss := atomic.LoadUint32(&store2.miss); miss != 1 {
 			t.Fatalf("unexpected miss count %v", miss)
 		}
-		if hits := atomic.LoadUint64(&store2.hits); hits != uint64(count-1) {
+		if hits := atomic.LoadUint32(&store2.hits); hits != uint32(count-1) {
 			t.Fatalf("unexpected hits count %v", hits)
 		}
 	})

--- a/pubsub.go
+++ b/pubsub.go
@@ -42,9 +42,9 @@ func newSubs() *subs {
 }
 
 type subs struct {
+	cnt uint64
 	chs map[string]chs
 	sub map[uint64]*sub
-	cnt uint64
 	mu  sync.RWMutex
 }
 

--- a/ring.go
+++ b/ring.go
@@ -3,6 +3,8 @@ package rueidis
 import (
 	"sync"
 	"sync/atomic"
+
+	"golang.org/x/sys/cpu"
 )
 
 type queue interface {
@@ -20,7 +22,7 @@ func newRing(factor int) *ring {
 		factor = DefaultRingScale
 	}
 	r := &ring{store: make([]node, 2<<(factor-1))}
-	r.mask = uint64(len(r.store) - 1)
+	r.mask = uint32(len(r.store) - 1)
 	for i := range r.store {
 		m := &sync.Mutex{}
 		r.store[i].c1 = sync.NewCond(m)
@@ -32,12 +34,12 @@ func newRing(factor int) *ring {
 
 type ring struct {
 	store []node // store's size must be 2^N to work with the mask
-	_     [5]uint64
-	write uint64
-	_     [7]uint64
-	read1 uint64
-	read2 uint64
-	mask  uint64
+	_     cpu.CacheLinePad
+	write uint32
+	_     cpu.CacheLinePad
+	read1 uint32
+	read2 uint32
+	mask  uint32
 }
 
 type node struct {
@@ -52,7 +54,7 @@ type node struct {
 }
 
 func (r *ring) PutOne(m Completed) chan RedisResult {
-	n := &r.store[atomic.AddUint64(&r.write, 1)&r.mask]
+	n := &r.store[atomic.AddUint32(&r.write, 1)&r.mask]
 	n.c1.L.Lock()
 	for n.mark != 0 {
 		n.c1.Wait()
@@ -68,7 +70,7 @@ func (r *ring) PutOne(m Completed) chan RedisResult {
 }
 
 func (r *ring) PutMulti(m []Completed, resps []RedisResult) chan RedisResult {
-	n := &r.store[atomic.AddUint64(&r.write, 1)&r.mask]
+	n := &r.store[atomic.AddUint32(&r.write, 1)&r.mask]
 	n.c1.L.Lock()
 	for n.mark != 0 {
 		n.c1.Wait()

